api/lang/symbols.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2015 ThoughtWorks, Inc.
+
+// This file is part of Gauge.
+
+// Gauge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Gauge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Gauge.  If not, see <http://www.gnu.org/licenses/>.
+
+package lang
+
+import (
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/getgauge/gauge/gauge"
+	"github.com/getgauge/gauge/logger"
+	"github.com/getgauge/gauge/parser"
+	"github.com/getgauge/gauge/util"
+	"github.com/sourcegraph/go-langserver/pkg/lsp"
+	"github.com/sourcegraph/jsonrpc2"
+)
+
+func documentSymbols(req *jsonrpc2.Request) (interface{}, error) {
+	var params lsp.DocumentSymbolParams
+	var err error
+	if err = json.Unmarshal(*req.Params, &params); err != nil {
+		logger.APILog.Debugf("failed to parse request %s", err.Error())
+		return nil, err
+	}
+	file := util.ConvertURItoFilePath(params.TextDocument.URI)
+	content := getContent(params.TextDocument.URI)
+	if util.IsConcept(file) {
+		return getConceptSymbols(content, file), nil
+	}
+	spec, parseResult := new(parser.SpecParser).Parse(content, gauge.NewConceptDictionary(), file)
```

Comment:
Do we need to parse the file? Can we not get it from the cache?
NOTE : InCase of workspace symbols we are getting it from cache.

Reply:
`documentSymbols` are generated from buffer content, so need to be parsed again. (Since the document can be opened and edited, without saving).

If we enable autosave, we do not need this.